src/app/modules/spcp/sp.oidc.client.cache.ts

```diff
@@ -0,0 +1,233 @@
+import axios from 'axios'
+import { createPublicKey } from 'crypto'
+import jwkToPem from 'jwk-to-pem'
+import NodeCache from 'node-cache'
+import { BaseClient, Issuer } from 'openid-client'
+
+import { createLoggerWithLabel } from '../../config/logger'
+
+import { JwkError } from './sp.oidc.client.errors'
+import {
+  CryptoKeys,
+  PublicJwks,
+  Refresh,
+  SecretJwks,
+  SpOidcClientCacheConstructorParams,
+} from './sp.oidc.client.types'
+import {
+  isEC,
+  retryPromiseForever,
+  retryPromiseThreeAttempts,
+} from './sp.oidc.util'
+
+// Name of keys in cache
+const BASE_CLIENT_NAME = 'baseClient'
+const NDI_PUBLIC_KEY_NAME = 'ndiPublicKeys'
+const EXPIRY_NAME = 'expiry'
+
+const logger = createLoggerWithLabel(module)
+
+/**
+ * Cache class which provides read-through capability and refresh-ahead before expiry
+ * Handles discovery and retrieval of NDI's public jwks
+ */
+export class SpOidcClientCache {
+  /**
+   * Cache to store NDI's keys and client
+   * @private
+   * accessible only for testing
+   */
+  _cache: NodeCache
+
+  /**
+   * Stores the refresh promise so that there is at most
+   * one in-flight refresh at any point in time.
+   * Stored promise is always pending
+   * @private
+   * accessible only for testing
+   */
+  _refreshPromise?: Promise<Refresh>
+
+  #spOidcNdiDiscoveryEndpoint: string
+  #spOidcNdiJwksEndpoint: string
+  #spOidcRpClientId: string
+  #spOidcRpRedirectUrl: string
+  #spOidcRpSecretJwks: SecretJwks
+
+  /**
+   * Constructor for cache
+   * On instantiation, trigger a refresh to populate the cache
+   */
+  constructor({
+    options,
+    spOidcNdiDiscoveryEndpoint,
+    spOidcNdiJwksEndpoint,
+    spOidcRpClientId,
+    spOidcRpRedirectUrl,
+    spOidcRpSecretJwks,
+  }: SpOidcClientCacheConstructorParams) {
+    this.#spOidcNdiDiscoveryEndpoint = spOidcNdiDiscoveryEndpoint
+    this.#spOidcNdiJwksEndpoint = spOidcNdiJwksEndpoint
+    this.#spOidcRpClientId = spOidcRpClientId
+    this.#spOidcRpRedirectUrl = spOidcRpRedirectUrl
+    this.#spOidcRpSecretJwks = spOidcRpSecretJwks
+    this._cache = new NodeCache(options)
+
+    // On expiry, refresh cache. If fail to refresh, log but do not throw error.
+    this._cache.on('expired', () =>
+      this.refresh().catch((err) =>
```

Comment:
Just checking: `refresh()` no longer has a fail mode, so the `catch()` here and on line 90 will never run, is that correct?